chronos/timer.nim

```diff
@@ -8,45 +8,397 @@
+  result.value = cast[int64](v) * Hour.value
+
+func days*(v: SomeInteger): Duration {.inline.} =
+  ## Initialize Duration with days value ``v``.
+  result.value = cast[int64](v) * Day.value
+
+func weeks*(v: SomeInteger): Duration {.inline.} =
+  ## Initialize Duration with weeks value ``v``.
+  result.value = cast[int64](v) * Week.value
+
+func nanoseconds*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to nanoseconds.
+  result = v
+
+func microseconds*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to microseconds.
+  result = (v div Microsecond) * Microsecond
+
+func milliseconds*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to milliseconds.
+  result = (v div Millisecond) * Millisecond
+
+func seconds*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to seconds.
+  result = (v div Second) * Second
+
+func minutes*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to minutes.
+  result = (v div Minute) * Minute
+
+func hours*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to hours.
+  result = (v div Hour) * Hour
+
+func days*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to days.
+  result = (v div Day) * Day
+
+func weeks*(v: Duration): Duration {.inline.} =
+  ## Round Duration ``v`` to weeks.
+  result = (v div Week) * Week
+
+func nanos*[T: SomeInteger|Duration](v: T): Duration {.inline.} =
+  nanoseconds(v)
+
+func micros*[T: SomeInteger|Duration](v: T): Duration {.inline.} =
+  microseconds(v)
+
+func millis*[T: SomeInteger|Duration](v: T): Duration {.inline.} =
+  milliseconds(v)
+
+func secs*[T: SomeInteger|Duration](v: T): Duration {.inline.} =
+  seconds(v)
+
+func toInt*(a: Duration, unit: Duration): int64 {.inline.} =
+  ## Round Duration ``a`` to specific Duration ``unit`` and returns value
+  ## as integer.
+  result = (a div unit).value
+
+func `$`*(a: Duration): string {.inline.} =
+  ## Returns string representation of Duration ``a`` as nanoseconds value.
+  result = $(a.value)
+
+func `$`*(a: Moment): string {.inline.} =
+  ## Returns string representation of Moment ``a`` as nanoseconds value.
+  result = $(a.value)
+
+func isOver*(a: Duration): bool {.inline.} =
+  ## Returns ``true`` if Duration ``a`` is ``0``.
+  result = (a.value == 0)
+
+func isInfinite*(a: Duration): bool {.inline.} =
+  ## Returns ``true`` if Duration ``a`` is infinite.
+  result = (a.value == -1)
+
+proc now*(t: typedesc[Moment]): Moment {.inline.} =
+  ## Returns current moment in time as Moment.
+  result.value = cast[int64](fastEpochTimeNano())
+
+proc init*(t: typedesc[Moment], value: int64, precision: Duration): Moment =
+  ## Initialize Moment with absolute time value ``value`` with precision
+  ## ``precision``.
+  result.value = value * precision.value
+
+proc fromNow*(t: typedesc[Duration], a: Duration): Moment {.inline.} =
+  result = Moment.now() + a
+
+func getTimestamp*(a: Duration): int {.inline.} =
```

Comment:
this looks a bit strange - ie, there's no standard that says something called a timestamp should be in `milliseconds` specifically (it's pretty common with micros these days, or nanos). I think it might be better not to have this function here, and instead use `nanoseconds` and do the rounding at the call site.

the easier non-branching way to do strict round up is `(a.value + 1000000-1) div 1000000`

further, there's no guarantee the resulting millisecond value will fit, needs 41 bits:
```
>>> math.log(2**63 / 1000000, 2)
43.06843143067571
```

Reply:
it looks like the template where getTimestamp is used could be rewritten to use durations throughout.. then in the windows section, it could convert to dword accordingly.. 

also, instead of rounding up, an alternative is to use `min(1, timeout)`